palace/fem/fespace.hpp

```diff
@@ -35,6 +38,167 @@ class FiniteElementSpace : public mfem::ParFiniteElementSpace
+                "Out of bounds request for finite element space at level " << l << "!");
+    return *fespaces[l];
+  }
+  const FiniteElementSpace &GetFESpaceAtLevel(std::size_t l) const
+  {
+    MFEM_ASSERT(l >= 0 && l < GetNumLevels(),
+                "Out of bounds request for finite element space at level " << l << "!");
+    return *fespaces[l];
+  }
+
+  FiniteElementSpace &GetFinestFESpace()
+  {
+    MFEM_ASSERT(!empty(), "Out of bounds request for finite element space at level 0!");
+    return *fespaces.back();
+  }
+  const FiniteElementSpace &GetFinestFESpace() const
+  {
+    MFEM_ASSERT(!empty(), "Out of bounds request for finite element space at level 0!");
+    return *fespaces.back();
+  }
+
+  const Operator &GetProlongationAtLevel(std::size_t l) const
+  {
+    MFEM_ASSERT(l >= 0 && l < GetNumLevels() - 1,
+                "Out of bounds request for finite element space prolongation at level "
+                    << l << "!");
+    return P[l] ? *P[l] : BuildProlongationAtLevel(l);
+  }
+
+  std::vector<const Operator *> GetProlongationOperators() const
+  {
+    std::vector<const Operator *> P_(GetNumLevels() - 1);
+    for (std::size_t l = 0; l < P_.size(); l++)
+    {
+      P_[l] = &GetProlongationAtLevel(l);
+    }
+    return P_;
+  }
+};
+
+//
+// A special type of FiniteElementSpaceHierarchy where all members are auxiliary finite
+// element spaces.
+//
+class AuxiliaryFiniteElementSpaceHierarchy : public FiniteElementSpaceHierarchy
+{
+public:
+  AuxiliaryFiniteElementSpaceHierarchy() = default;
+  AuxiliaryFiniteElementSpaceHierarchy(
+      std::unique_ptr<AuxiliaryFiniteElementSpace> &&fespace)
+  {
+    AddLevel(std::move(fespace));
+  }
+
+  void AddLevel(std::unique_ptr<FiniteElementSpace> &&fespace) override
+  {
+    MFEM_ABORT("All finite element spaces in an AuxiliaryFiniteElementSpaceHierarchy must "
+               "inherit from AuxiliaryFiniteElementSpace!");
+  }
+
+  void AddLevel(std::unique_ptr<AuxiliaryFiniteElementSpace> &&fespace)
+  {
+    // Guarantees that every object in fespaces is an AuxiliaryFiniteElementSpace.
+    fespaces.push_back(std::move(fespace));
+    P.push_back(nullptr);
+  }
+
+  AuxiliaryFiniteElementSpace &GetAuxiliaryFESpaceAtLevel(std::size_t l)
+  {
+    return *static_cast<AuxiliaryFiniteElementSpace *>(&GetFESpaceAtLevel(l));
+  }
+  const AuxiliaryFiniteElementSpace &GetAuxiliaryFESpaceAtLevel(std::size_t l) const
+  {
+    return *static_cast<const AuxiliaryFiniteElementSpace *>(&GetFESpaceAtLevel(l));
+  }
+
+  AuxiliaryFiniteElementSpace &GetFinestAuxiliaryFESpace()
+  {
+    return *static_cast<AuxiliaryFiniteElementSpace *>(&GetFinestFESpace());
+  }
+  const AuxiliaryFiniteElementSpace &GetFinestAuxiliaryFESpace() const
+  {
+    return *static_cast<const AuxiliaryFiniteElementSpace *>(&GetFinestFESpace());
+  }
+
+  const Operator &GetDiscreteInterpolatorAtLevel(std::size_t l) const
+  {
+    return GetAuxiliaryFESpaceAtLevel(l).GetDiscreteInterpolator();
+  }
```

Comment:
`GetDiscreteInterpolatorAtFinestLevel` for consistency with the other getters.

Reply:
There is no `GetProlongationAtFinestLevel` so I have chosen to not include a `GetDiscreteInterpolatorAtFinestLevel`.